lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp

```diff
@@ -550,6 +550,26 @@ static bool RemoveUnstructuredLoopExitsIteration(BasicBlock *exiting_block,
   }
 
   BasicBlock *latch_exit = GetExitBlockForExitingBlock(L, latch);
+
+  // Ensure the latch-exit is "dedicated": no block outside the loop
+  // branches to it.
+  //
+  // Suppose this iteration successfully moves an exit block X until
+  // after the latch block.  It will do so by rewiring the CFG so
+  // the latch *exit* block will branch to X.  If the latch exit
+  // block is already reachable from X, then the rewiring will
+  // create an unwanted loop.
+  // So prevent this from happening by ensuring the latch exit is
+  // "dedicated": the only branches to it come from inside the
+  // loop, and hence not from X.
+  for (auto *pred : predecessors(latch_exit)) {
```

Comment:
I was wondering about the case where the `latch_exit` has multiple precessors outside the loop (if that is even possible) and if we needed to split those edges as well.

But then I realized that this is splitting the edge between the `latch` and the `latch_exit` and the new block on the edge will become the new `latch_exit` when we recalculate the exit blocks. Is that correct? I think it wouldn't hurt to mention in the comment if so.

Reply:
+1   This code does handle that case: latch-exit block has multiple predecessors from outside the loop.

You're correct, once the edge is split we have this picture:

       latch --> middle --> tail

The 'middle' block is the new latch exit block; its only predecessor is 'latch'.

The edges (from blocks outside the loop) that used to go to latch exit block now go to tail.

Reply:
> I was wondering about the case where the latch_exit has multiple precessors outside the loop (if that is even possible) 

That depends on the situations that this routine is called in.  I don't know if the standard flow can make this happen.
I hit this case when I was making manual test cases.

Reply:
Added a comment as requested.

Reply:
Also, the 'shared-exit.ll' test exercises the case you were wondering about.